File: social/images.py
"""Ask Shopify's image CDN for a card-sized copy of a product photo.

A direct port of ``shopImage`` in frontend/src/lib/images.js, and deliberately
the same fifteen lines rather than anything cleverer: the frontend and the post
card should ask the CDN for images the same way, and the reason for the
single-host scoping is unchanged. 98% of our product images are on
cdn.shopify.com and none of the scraped URLs carry a size, so without this the
renderer pulls a 2-3 MB original to paint a 1000px slot. The other ~2% sit on
assorted vendor CDNs with no shared transform contract and pass through
untouched.
"""
from urllib.parse import urlencode, urlparse, urlunparse, parse_qsl

SHOPIFY_CDN_HOST = "cdn.shopify.com"


def shop_image(url: str | None, width: int) -> str | None:
    if not url:
        return url
    try:
        parts = urlparse(url)
        if parts.hostname != SHOPIFY_CDN_HOST:
            return url
        # Overwrite any width already present rather than appending a second
        # one, and preserve Shopify's ?v= cache key.
        query = [(k, v) for k, v in parse_qsl(parts.query) if k != "width"]
        query.append(("width", str(width)))
        return urlunparse(parts._replace(query=urlencode(query)))
    except ValueError:
        return url

File: frontend/src/lib/images.js
// Ask Shopify's image CDN for a display-sized copy instead of the full-res
// original.
//
// 98% of our product images are on cdn.shopify.com, and none of the scraped
// URLs carry a size, so every card was downloading a 1-3 MB photo to paint a
// slot a few hundred pixels wide. The CDN resizes on the fly from a `width`
// query parameter: a 2.3 MB hero comes back at ~170 KB for a 600px card and
// ~10 KB for a 120px table thumbnail, keeping the aspect ratio and returning
// JPEG.
//
// Only cdn.shopify.com is rewritten. The other ~2% of images sit on assorted
// vendor CDNs with no shared transform contract, so those pass through
// untouched. Anything that is not a valid http(s) URL is returned unchanged;
// callers still guard with isHttpUrl before rendering.
//
// Import-free (no JSX, no other modules) so the node --test runner can load it
// directly, the same arrangement as the other lib/ suites.

export function shopImage(url, width) {
  if (!url) return url
  try {
    const u = new URL(url)
    // Scoped to the one host whose transform contract we know. searchParams.set
    // handles both a bare URL and one that already carries Shopify's ?v= cache
    // key, and overwrites any width already present rather than appending a
    // second one.
    if (u.hostname !== 'cdn.shopify.com') return url
    u.searchParams.set('width', String(width))
    return u.toString()
  } catch {
    return url
  }
}
